clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -350,27 +352,39 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
         if (numHeaders == 0)
             return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
 
+        Header[] headers = getHeaders(buffer, numHeaders, recordStart, sizeInBytes);
+
+        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+    }
+
+    private static Header[] getHeaders(ByteBuffer buffer, int numHeaders, int recordStart, int sizeInBytes) {
         Header[] headers = new Header[numHeaders];
-        for (int i = 0; i < numHeaders; i++) {
-            int headerKeySize = ByteUtils.readVarint(buffer);
-            if (headerKeySize < 0)
-                throw new InvalidRecordException("Invalid negative header key size " + headerKeySize);
-
-            String headerKey = Utils.utf8(buffer, headerKeySize);
-            buffer.position(buffer.position() + headerKeySize);
-
-            ByteBuffer headerValue = null;
-            int headerValueSize = ByteUtils.readVarint(buffer);
-            if (headerValueSize >= 0) {
-                headerValue = buffer.slice();
-                headerValue.limit(headerValueSize);
-                buffer.position(buffer.position() + headerValueSize);
+        try {
+            for (int i = 0; i < numHeaders; i++) {
+                int headerKeySize = ByteUtils.readVarint(buffer);
+                if (headerKeySize < 0)
+                    throw new InvalidRecordException("Invalid negative header key size " + headerKeySize);
+
+                String headerKey = Utils.utf8(buffer, headerKeySize);
+                buffer.position(buffer.position() + headerKeySize);
+
+                ByteBuffer headerValue = null;
+                int headerValueSize = ByteUtils.readVarint(buffer);
+                if (headerValueSize >= 0) {
+                    headerValue = buffer.slice();
+                    headerValue.limit(headerValueSize);
+                    buffer.position(buffer.position() + headerValueSize);
+                }
+
+                headers[i] = new RecordHeader(headerKey, headerValue);
             }
-
-            headers[i] = new RecordHeader(headerKey, headerValue);
+        } catch (BufferUnderflowException e) {
+            throw new InvalidRecordException("Invalid header data or number of headers declared for the record");
         }
-
-        return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, headers);
+        // validate whether we have read all header bytes in the current record
+        if (buffer.position() - recordStart != sizeInBytes - ByteUtils.sizeOfVarint(sizeInBytes))
```

Comment:
We are already passing it. `sizeInBytes` is `ByteUtils.sizeOfVarint(sizeOfBodyInBytes) + sizeOfBodyInBytes` (passed into `readFrom`). In this check we are trying to find out whether actually `sizeOfBodyInBytes` (minus `sizeOfVarint(sizeOfBodyInBytes)` to account for this count itself) bytes have been read from the `buffer` in this record.